tests/internal/test_module.py

```diff
@@ -0,0 +1,254 @@
+from os.path import dirname
+import sys
+
+import mock
+import pytest
+
+from ddtrace.internal.module import ModuleWatchdog
+from ddtrace.internal.module import origin
+import tests.test_module
+
+
+@pytest.fixture
+def module_watchdog():
+    ModuleWatchdog.install()
+
+    assert ModuleWatchdog.is_installed()
+
+    yield ModuleWatchdog
+
+    ModuleWatchdog.uninstall()
+
+
+def test_watchdog_install_uninstall():
+    assert not isinstance(sys.modules, ModuleWatchdog)
+    ModuleWatchdog.install()
+    assert isinstance(sys.modules, ModuleWatchdog)
+    ModuleWatchdog.uninstall()
+    assert not isinstance(sys.modules, ModuleWatchdog)
+
+
+def test_import_origin_hook_for_imported_module(module_watchdog):
+    hook = mock.Mock()
+    module = sys.modules[__name__]
+    module_watchdog.register_origin_hook(origin(module), hook)
+
+    hook.assert_called_once_with(module)
+
+
+def test_import_module_hook_for_imported_module(module_watchdog):
+    hook = mock.Mock()
+    module = sys.modules[__name__]
+    module_watchdog.register_module_hook(module.__name__, hook)
+
+    hook.assert_called_once_with(module)
+
+
+def test_register_hook_without_install():
+    with pytest.raises(RuntimeError):
+        ModuleWatchdog.register_origin_hook(__file__, mock.Mock())
+
+    with pytest.raises(RuntimeError):
+        ModuleWatchdog.register_module_hook(__name__, mock.Mock())
+
+
+@pytest.mark.subprocess(env=dict(MODULE_ORIGIN=origin(tests.test_module)))
+def test_import_origin_hook_for_module_not_yet_imported():
+    import os
+    import sys
+
+    from mock import mock
+
+    from ddtrace.internal.module import ModuleWatchdog
+
+    name = "tests.test_module"
+    path = os.getenv("MODULE_ORIGIN")
+    hook = mock.Mock()
+
+    ModuleWatchdog.install()
+
+    ModuleWatchdog.register_origin_hook(path, hook)
+
+    hook.assert_not_called()
```

Comment:
Can you add an import before this as the register should not call the hook?

Reply:
We already have some modules loaded at this stage so this should already show that we are not triggering the hook at the wrong time. Instead, I have added another import below to check that we don't fire on the wrong module import